pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+                    " out of buffer {end_bytes}."\
+                    "".format(cursor=self.cursor,
+                              end_bytes=self.dmap_end_bytes)
+            raise pydmap_exceptions.CursorError(self.cursor, message=message)
+
+        cursor_offset = self.dmap_end_bytes - self.cursor
+        if cursor_offset < data_fmt_bytes:
+            message = "Error: Byte offsets {offset} into buffer {fmt}"\
+                    " are not aligned.".format(offset=cursor_offset,
+                                               fmt=data_fmt_bytes)
+            raise pydmap_exceptions.CursorError(self.cursor, message=message)
+
+        # struct.unpack is a python method the converts C struct as python bytes
+        # to python values. See DEVELOPER_README.md for more information.
+        # TODO: how to handle this gracefully? Return None and check later than
+        #       continue in the loop?
+        if data_type_fmt == 'c':
+            data = self.dmap_bytearr[self.cursor]
+            self.cursor += data_fmt_bytes
+            return data
+        elif data_type_fmt == 's':
+            byte_counter = 0
+            while self.dmap_bytearr[self.cursor + byte_counter] != 0 or self.cursor + byte_counter >= self.dmap_end_bytes:
+                # add 1 byte to byte_counter because a string is a list of chars
+                byte_counter += 1  # 1 is the number of bytes of a char
+            self.bytes_check(self.cursor+byte_counter, 'cursor', self.dmap_end_bytes, 'Total number of Bytes')
+
+            # format byte counter to a string format for unpacking
+            # WARNING: not using format for performance purposes
+            char_count = '%ss'%byte_counter
+            data = struct.unpack_from(char_count,
+                                      self.dmap_buffer,
+                                      self.cursor)
+            ## 1 is the number of bytes in a char
+            self.cursor = self.cursor + byte_counter + 1
+            return data[0].decode('utf-8')
+        else:
+            data = struct.unpack_from(data_type_fmt,
+                                      self.dmap_buffer,
+                                      self.cursor)
+            self.cursor += data_fmt_bytes
+            return data[0]
+
+    def read_string_array(self, shape: list,
+                          array_type_fmt: str,
+                          array_fmt_bytes: int) -> np.ndarray:
+        """
+        Reads and builds a N-D array of string data types.
+
+        Parameter
+        ---------
+        shape : list
+            list of the arrays dimension size
+        array_fmt_bytes : int
+            String or char
+
+        Return
+        ------
+        np.ndarray
+            an numpy N-D array of string read in from the byte array
+        """
+
+        data = []
+        # FIXME: still need to find a method for testing this... Or a file
+        #        that contains a string/DMAP array
+        for dim_size in shape:
```

Comment:
I don't think this is working correctly. if shape is (3,3,3) for example your final array is going to have 9 elements, whereas your 3x3x3 array should have 27 elements. I think this is why I used recursion for this method. This is also going to leave your cursor in the wrong spot. You are also appending these values to a flat array without reshaping before the return.

Reply:
Thus the FIXME comment. Once I have DmapWrite working, I can do some integration testing and try out an array of strings to figure out the loop structures. 

Currently, there is not a DMAP file type that has an array of strings, so I was unable to test this function correctly.